### DISCUSSION TITLE
User experience with use `async_provide_credentials_fn` and `LazyCachingCredentialsProvider`

### DISCUSSION THREAD

**Discussion by Eh2406** [awslabs/aws-sdk-rust#174] [General]
I had a chance to try `async_provide_credentials_fn` and `LazyCachingCredentialsProvider`. I may have bean using them wrong, but here is how it went for me. My use case looked like:
```rs
async fn get_creds_form_outside_prosses(acc: &str) -> aws_auth::provider::CredentialsResult {
   ...
}
...
for account in accounts {
    let creds = get_creds_form_outside_prosses(&account);
    ....
    let db_config = dynamodb::Config::builder()
        .credentials_provider(
            creds.clone() // the clone is needed because I will also need them for `cw_config`
        ).build();
    let db_client = dynamodb::Client::from_conf(db_config);
    ....
    let cw_config = cloudwatchlogs::Config::builder()
        .credentials_provider(creds)
        .build();
    let cw_client = cloudwatchlogs::Client::from_conf(cw_config);
    ...
}

```

Now when I am typing this out, I am realizing that somehow (I don't see how) `Credentials` from a `CredentialsResult` impls `AsyncProvideCredentials`. So using it directly works, and is much cleaner then what I tried before! 
- Lesson 1: how can we make that more obvious when navigating the code?

Yesterday, I started by looking at `AsyncProvideCredentials` witch is impled by `async_provide_credentials_fn` witch has an example in the doc string, so let me try using that.
```rs
for account in accounts {
    let creds = async_provide_credentials_fn(|| get_creds_form_outside_prosses(&account));
    ....
```
```
error[E0597]: `account` does not live long enough
...
returning this value requires that `account` is borrowed for `'static`
- `account` dropped here (at end of loop) while still borrowed
```
Why `'static`? because `.credentials_provider(` requires it `impl aws_auth::provider::AsyncProvideCredentials + 'static`.
I know the arcane trick to get a `&'static str`! Then I end up needing to add a `move`
```rs
for account in accounts {
    let account = &*Box::leak(account.into_boxed_str());
    let creds = async_provide_credentials_fn(move || get_creds_form_outside_prosses(&account));
    ....
```
And this is a working solution!
- Lesson 2: Why `'static`? How can we make this better?

Some one told me to try `LazyCachingCredentialsProvider`. So:
```rs
for account in accounts {
    let account = &*Box::leak(account.into_boxed_str());
    let creds = LazyCachingCredentialsProvider::builder()
       .refresh(async_provide_credentials_fn(|| get_creds_form_outside_prosses(&account)))
       .build();
    ....
```
but
```
no method named `clone` found for struct `LazyCachingCredentialsProvider` in the current scope
```
- Lesson 3: How do I use `LazyCachingCredentialsProvider` in more then one `config`?
- Lesson 4: Is there some way to make a `config` or `client` that can be used for different services?

**Comment by rcoh** [comment]
this extremely helpful. thanks for taking the time to write this up. We're going to take another pass on both the trait as well as write a usage guide: #175.

And for your and other's future context, `Credentials` implements `AsyncProvideCredentials` because of a blanket impl on `impl AsyncProvideCredentials for T where T: ProvideCredentials`

**Comment by carllerche** [comment]
I'm not sure why you need a `Box::leak` here. It is the closure that needs to be `'static` so it should be sufficient to move an owned `String` into the closure.

```rust
let account = account.to_string();
let creds = async_provide_credentials_fn(move || get_creds_form_outside_prosses(&account));
```

**Reply by Eh2406** [reply]
I did try that before posting, and @rcoh tried after I posted, and we could not get it to work. But we may have bean messing things up.

**Reply by jdisanti** [reply]
I tried it out, and the only way I could get it to compile was to change `get_creds_from_outside_process` to take the input as an owned `String` and to also clone it (it complains about not being able to copy String otherwise):
```rust
async fn get_creds_from_outside_process(input: String) -> aws_auth::provider::CredentialsResult {
    unimplemented!()
}

fn test() {
    let accounts = vec!["one".to_string(), "two".to_string()];
    for account in accounts {
        let creds =
            async_provide_credentials_fn(move || get_creds_from_outside_process(account.clone()));
    }
}
```
Any attempt at borrowing the string value resulted in it not living long enough, no matter what permutation I tried.
Definitely seems sub-optimal.
